### PR TITLE
Add ParoQuant (pairwise rotation quantization) support

### LINKED ISSUE #977: Add ParoQuant (Pairwise Rotation Quantization) Support to MLX-LM
### Feature Request
I would like to request support for ParoQuant (Pairwise Rotation Quantization) in MLX-LM.

### Background
ParoQuant is a new post-training quantization method introduced at ICLR 2026. It uses pairwise rotation to suppress outliers in weight distributions, making it especially effective for reasoning-heavy LLMs. Compared to standard PTQ, it achieves better accuracy retention while still reducing memory and compute costs.

### Why MLX-LM?
MLX-LM already supports uniform, mixed-bit, and affine quantization. Adding ParoQuant would:
- Improve robustness for reasoning-focused models.
- Enable developers to experiment with cutting-edge quantization methods on Apple Silicon.
- Keep MLX aligned with the latest research in efficient LLM deployment.

### Suggested Implementation
- Integrate ParoQuant’s pairwise rotation preprocessing step into MLX’s quantization pipeline.
- Provide options for INT4/INT8 precision.
- Allow exporting/importing ParoQuant-quantized weights for Hugging Face compatibility.

### References
- ParoQuant paper (ICLR 2026): Liang, Chen, Han, Liu
- GitHub repo: z-lab/paroquant

Thanks for considering this request!

### PR DESCRIPTION
**Pull request by Ziqiao-git**
# Add ParoQuant (pairwise rotation quantization) support

Closes #977.

## Summary

Adds inference support for **ParoQuant** (Pairwise Rotation Quantization,
[ICLR 2026, arXiv:2511.10645](https://arxiv.org/abs/2511.10645)). ParoQuant
applies a learned sequence of pairwise Givens rotations to the activations
*before* an affine-quantized matmul; the rotation decorrelates channels so 4-bit
quantization keeps more of the signal, at near-AWQ speed.

Loading a public ParoQuant checkpoint now works with no extra arguments:

```python
from mlx_lm import load, generate

model, tok = load("z-lab/Meta-Llama-3-8B-PARO")   # auto-detected as paroquant
print(generate(model, tok, "The capital of France is", max_tokens=40))
```

`generate` / `chat` / `server` all work out of the box.

## What's added

- **`mlx_lm/models/paroquant.py`** — the inference layers and the rotation kernel:
  - `RotateQuantizedLinear`: `rotate(x)` (Metal kernel) → `mx.quantized_matmul`.
  - `RotateSwitchGLU`: a `SwitchGLU` with a shared rotation before gate/up and
    before down, wrapping the standard `QuantizedSwitchLinear` experts.
  - A JIT-compiled, cached `mx.fast.metal_kernel` for the pairwise rotation
    (inline source, in the style of `ssm.py` / `gated_delta.py`), with a guard
    that raises if a checkpoint's `krot`/`group_size` exceeds the kernel's
    compile-time limits.
  - The layers `freeze()` like `nn.QuantizedLinear` / `QuantizedSwitchLinear`, so
    the tuner/LoRA treats them as ordinary quantized layers.

- **`mlx_lm/utils.py`** — a `quant_method == "paroquant"` branch in `load_model`,
  parallel to the AWQ branch:
  - Public z-lab checkpoints ship **AutoAWQ-packed** weights
    (`qweight`/`qzeros`/`scales`), so weight conversion **reuses the existing
    `_transform_awq_weights`** (bit-exact vs. the reference converter); rotation
    params pass through.
  - `_patch_paro_layers` swaps `nn.Linear` (with a `.theta`) → `RotateQuantizedLinear`,
    `SwitchLinear` → `QuantizedSwitchLinear`, and `SwitchGLU` (with shared
    rotations) → `RotateSwitchGLU` before `load_weights`.
  - Per-expert MoE weights are stacked into the `switch_mlp` layout and shared
    rotations are remapped to `gate_up_rot` / `down_rot`.
  - `embed_tokens` / `lm_head` are kept in their checkpoint (float) form.

## How this differs from AWQ

The integration point and the weight *layout* conversion mirror AWQ (which is why
`_transform_awq_weights` is reused, bit-exact). But ParoQuant is a different
quantization *method*: the pairwise rotation has no AWQ counterpart, so it needs a
new layer (`RotateQuantizedLinear`) and the Metal kernel rather than loading as a
plain `QuantizedLinear`.

## Testing

`tests/test_paroquant.py` (unit):
- rotation kernel vs. a NumPy reference, for both fixed and random within-group
  matchings;
- AWQ-packed weight transform shapes / dtypes / rotation pass-through;
- dense and MoE layer patching;
- a MoE forward smoke (`_stack_paro_moe_experts` → `QuantizedSwitchLinear` →
  `RotateSwitchGLU` → finite output).

End-to-end on real public z-lab checkpoints (coherent generation, no NaNs):

| Model | Architecture |
|---|---|
| `z-lab/Meta-Llama-3-8B-PARO` | dense attention |
| `z-lab/Llama-3.1-8B-Instruct-PARO` | dense attention (chat template) |
| `z-lab/Qwen3.5-2B-PARO` | dense linear-attention (gated delta) |
| `z-lab/Qwen3.5-35B-A3B-PARO` | MoE + gated delta (multimodal text path) |
| `z-lab/gemma-4-31B-it-PARO` | Gemma 4 (per-layer embeddings, multimodal text path) |

Weight conversion was checked bit-exact against ParoQuant's own reference
converter on real Llama-8B weights (`max|diff| = 0` for weight/scales/biases).

## Notes

- No changes to the mlx framework or to `convert.py`; ParoQuant models are
  produced by ParoQuant's own tooling, and mlx-lm only loads and runs them.
- 4-bit only (matches the published checkpoints and the reused AWQ path).
- Instruction-tuned checkpoints need their chat template to generate coherently
  (e.g. Gemma 4 produces repetition on a raw prompt, as noted in #1123); this is
  not specific to ParoQuant.

### CODE REVIEW

**Comment by zcbenz** [comment]
Thanks for the PR but at the moment the number of PRs is way beyond our capacity to review so I'm closing the non-essential ones so we can actually work on this repo.
